turbopack/crates/turbopack-ecmascript/src/tree_shake/optimizations.rs

```diff
@@ -0,0 +1,110 @@
+    where
+        N: Copy,
+        Self: Index<N, Output = ItemId>,
+    {
+        let item_id = &self[*item];
+
+        matches!(
+            item_id,
+            ItemId::Item {
+                kind: ItemIdItemKind::ImportBinding(..),
+                ..
+            }
+        )
+    }
+    pub(super) fn should_not_merge_iter<N>(&self, items: &[N]) -> bool
+    where
+        N: Copy,
+        Self: Index<N, Output = ItemId>,
+    {
+        items.iter().any(|item| self.should_not_merge(item))
+    }
+
+    /// Optimizes a condensed graph by merging nodes with only one incoming edge.
+    pub(super) fn merge_single_incoming_nodes<N>(&self, g: &mut Graph<Vec<N>, Dependency>)
+    where
+        N: Copy,
+        Self: Index<N, Output = ItemId>,
+    {
+        let mut queue = vec![];
+        let mut removed_nodes = vec![];
+
+        for node in g.node_indices() {
+            // ImportBinding nodes should not be merged
+            let node_data = g.node_weight(node).expect("Node should exist");
+            if self.should_not_merge_iter(node_data) {
+                continue;
+            }
+
+            // If the node has only one incoming edge, we enqueue it
+            if g.edges_directed(node, Direction::Incoming).count() == 1 {
+                let dependant = g
+                    .edges_directed(node, Direction::Incoming)
+                    .next()
+                    .unwrap()
+                    .source();
+
+                if self.should_not_merge_iter(&g[dependant]) {
+                    continue;
+                }
+
+                let dependencies = g
+                    .edges_directed(node, Direction::Outgoing)
+                    .map(|e| (e.target(), *e.weight()))
+                    .collect::<Vec<_>>();
+
+                queue.push((node, dependant, dependencies));
+                removed_nodes.push(node);
+            }
+        }
+
+        for (original, dependant, dependencies) in queue {
+            // Move all edges from node to dependant
+            for (dependency, weight) in dependencies {
+                let edge = g
+                    .find_edge(dependant, dependency)
+                    .and_then(|e| g.edge_weight_mut(e));
+                match edge {
+                    Some(v) => {
+                        if matches!(v, Dependency::Weak) {
+                            *v = weight;
+                        }
+                    }
+                    None => {
+                        g.add_edge(dependant, dependency, weight);
+                    }
+                }
+            }
+
+            // Move items from original to dependant
+            let items = g.node_weight(original).expect("Node should exist").clone();
+            g.node_weight_mut(dependant).unwrap().extend(items);
+        }
+
+        // Remove all edges from source
+        for node in removed_nodes.into_iter().rev() {
+            g.remove_node(node).expect("Node should exist");
```

Comment:
After merging two nodes this might create a new merge-able edge. You want to check the edges of the merged node again.

Reply:
Yeah... That's why I made this as a function, but I forgot to call the function again.